code/computational_geometry/src/sphere_tetrahedron_intersection/sphere_tetrahedron_intersection.cpp

```diff
@@ -0,0 +1,1237 @@
+
+	// LineNormal N=2
+	template< class T >
+	inline LuVector< 2, T > LineNormal( const LuVector< 2, T >& v1, const LuVector< 2, T >& v2 ) 
+	{
+		LuVector< 2, T > v12 = Unit( v2 - v1 );
+		return OrthogonalR( v12 );
+	}
+
+	// LineNormalL N=3
+	template< class T >
+	inline LuVector< 3, T > LineNormalL( const LuVector< 3, T >& lineDir ) 
+	{
+		return OrthogonalR( lineDir );
+	}
+
+	// LineNormalL N=2
+	template< class T >
+	inline LuVector< 2, T > LineNormalL( const LuVector< 2, T >& lineDir ) 
+	{
+		return OrthogonalR( lineDir );
+	}
+
+	// LineNormalP N=3
+	template< class T >
+	inline LuVector< 3, T > LineNormalP( const LuVector< 3, T >& vec, const LuVector< 3, T >& v1, const LuVector< 3, T >& v2 ) 
+	{
+		LuVector< 3, T > v12 = Unit( v2 - v1 );
+		LuVector< 3, T > ort = vec - ProjLineL( vec, v1, v12 );
+		T len = Length( ort );
+		return len > 0 ? ort / len : LineNormalL( v12 );
+	}
+
+	// LineNormalP N=2
+	template< class T >
+	inline LuVector< 2, T > LineNormalP( const LuVector< 2, T >& vec, const LuVector< 2, T >& v1, const LuVector< 2, T >& v2 ) 
+	{
+		LuVector< 2, T > v12 = Unit( v2 - v1 );
+		LuVector< 2, T > ort = vec - ProjLineL( vec, v1, v12 );
+		T len = Length( ort );
+		return len > 0 ? ort / len : LineNormalL( v12 );
+	}
+
+	// LineNormalPL N=3
+	template< class T >
+	inline LuVector< 3, T > LineNormalPL( const LuVector< 3, T >& vec, const LuVector< 3, T >& v1, const LuVector< 3, T >& lineDir )  
+	{
+		LuVector< 3, T > ort = vec - ProjLineL( vec, v1, lineDir );
+		T len = Length( ort );
+		return len > 0 ? ort / len : LineNormalL( lineDir );
+	}
+
+	// LineNormalPL N=2
+	template< class T >
+	inline LuVector< 2, T > LineNormalPL( const LuVector< 2, T >& vec, const LuVector< 2, T >& v1, const LuVector< 2, T >& lineDir ) 
+	{
+		LuVector< 2, T > ort = vec - ProjLineL( vec, v1, lineDir );
+		T len = Length( ort );
+		return len > 0 ? ort / len : LineNormalL( lineDir );
+	}
+
+	//////////////////////////////////////////////////////////////
+
+	// ProjPlane
+	template< class T >
+	inline LuVector< 3, T > ProjPlane( const LuVector< 3, T >& vec, const LuVector< 3, T >& v1, const LuVector< 3, T >& n ) 
+	{
+		return vec - Dot( vec - v1, n ) * n;
+	}
+
+	// PlaneNormal
+	template< class T >
+	inline LuVector< 3, T > PlaneNormal( const LuVector< 3, T >& v1, const LuVector< 3, T >& v2, const LuVector< 3, T >& v3 )
+	{
+		return Unit( Cross( v2 - v1, v3 - v1 ) );
+	}
+
+	// PlaneNormalP
+	template< class T >
+	inline LuVector< 3, T > PlaneNormalP(  const LuVector< 3, T >& vec, const LuVector< 3, T >& v1, const LuVector< 3, T >& v2, const LuVector< 3, T >& v3 ) 
+	{
+		LuVector< 3, T > n = PlaneNormal( v1, v2, v3 );
+		return Dot( n, vec - v1 ) > 0 ? n : -n;
+	}
+
+	// PlaneNormalPN
+	template< class T >
+	inline LuVector< 3, T > PlaneNormalPN(  const LuVector< 3, T >& vec, const LuVector< 3, T >& v1, const LuVector< 3, T >& n ) 
+	{
+		return Dot( n, vec - v1 ) > 0 ? n : -n;
+	}
+
+};
+#endif
+
+const double pi = 3.14159265358979323846;
```

Comment:
I'd prefer that you namespace this variable in something called `Constants` or whatnot. (And maybe mark it constexpr). This may reduce the chance for a naming conflict.

Reply:
Cancer of the C++ :)
Even more cancerous thing is defining it like "pi = atan(1.0)*4;" :D
https://stackoverflow.com/questions/1727881/how-to-use-the-pi-constant-in-c
I mean, do they even...?

I'm planning to refactor the intersection codes, I'll put it in a better place.
By the way, constexpr is not necessary if it's not an expression like "atan(1.0)*4", does it?

Reply:
Well you can't use it if it has a function call anyway, so I mean why not just write it out?